pennylane/labs/dla/cartan_subalgebra.py

```diff
@@ -0,0 +1,327 @@
+
+        >>> from pennylane.labs.dla import adjvec_to_op
+        >>> h = adjvec_to_op(np_h, g)
+        >>> h_op = [qml.pauli_decompose(op).pauli_rep for op in h]
+        >>> h_op
+        [0.35355339059327384 * Y(1) @ Y(2),
+         0.3535533905932738 * Y(0) @ Y(2),
+         -0.3535533905932738 * Y(0) @ Y(1)]
+
+        In that case we chose a Cartan subalgebra from which we can readily see that it is commuting, but we also provide a small helper function to check that.
+
+        >>> from pennylane.labs.dla import check_all_commuting
+        >>> assert check_all_commuting(h_op)
+
+        Last but not least, the adjoint representation ``new_adj`` is updated to represent the new basis and its ordering of ``g``.
+        In particular, we can compute commutators between
+    """
+
+    np_m = op_to_adjvec(m, g)
+    np_h = op_to_adjvec([m[start_idx]], g)
+
+    iteration = 1
+    while True:
+        if verbose:
+            print(f"iteration {iteration}: Found {len(np_h)} independent Abelian operators.")
+        kernel_intersection = np_m.T
+        for h_i in np_h:
+
+            # obtain adjoint rep of candidate h_i
+            adjoint_of_h_i = np.tensordot(ad, h_i, axes=[[1], [0]])
+            # compute kernel of adjoint
+            new_kernel = null_space(adjoint_of_h_i, rcond=tol)
+
+            # intersect kernel to stay in m
+            kernel_intersection = _intersect_bases(kernel_intersection, new_kernel, rcond=tol)
+
+        print(kernel_intersection.T.shape, np_h.shape)
+
+        if kernel_intersection.shape[1] == len(np_h):
+            # No new vector was added from all the kernels
+            break
+
+        kernel_intersection = _gram_schmidt(kernel_intersection)  # orthogonalize
+        for vec in kernel_intersection.T:
+            if _is_independent(vec, np.array(np_h).T, tol):
+                np_h = np.vstack([np_h, vec])
+                break
+
+        iteration += 1
+
+    np_h = _gram_schmidt(np_h.T).T  # orthogonalize Abelian subalgebra
+    np_k = op_to_adjvec(k, g)  # adjoint vectors of k space for re-ordering
+
+    np_mtilde = _orthogonal_complement_basis(np_h, np_m, tol=tol)  # the "rest" of m without h
+    np_newg = np.vstack([np_k, np_mtilde, np_h])
+
+    # Instead of recomputing the adjoint representation, take the basis transformation
+    # oldg -> newg and transform the adjoint representation accordingly
+    # TODO: implementation to be tested:
+    # basis change old g @ X = new g => adj_new = contact(adj, X, X, X)
+    basis_change = np.linalg.lstsq(np.vstack([np_k, np_m]).T, np_newg.T, rcond=None)[0]
+    # Perform the einsum contraction "kmn,ki,mj,nl->ijl" via three tensordot steps
+    new_adj = np.tensordot(ad, basis_change, axes=[[0], [0]])  # "kmn,ki->mni"
+    new_adj = np.tensordot(ad, basis_change, axes=[[0], [0]])  # "mni,mj->nij"
+    new_adj = np.tensordot(ad, basis_change, axes=[[0], [0]])  # "nij,nl->ijl"
```

Comment:
@dwierichs is this intentionally three times the same index that is contracted? Shouldnt this be 
```suggestion
    new_adj = np.tensordot(ad, basis_change, axes=[[0], [0]])  # "kmn,ki->mni"
    new_adj = np.tensordot(ad, basis_change, axes=[[1], [0]])  # "nmi,mj->nji"
    new_adj = np.tensordot(ad, basis_change, axes=[[2], [0]])  # "nij,jl->nil"
```

Reply:
e.g. this doesnt agree anymore:

```python
gens = [X(0) @ X(1), Z(0), Z(1)]
gens = [op.pauli_rep for op in gens]
g = qml.lie_closure(gens, pauli=True)

k, m = cartan_decomposition(g, even_odd_involution)
assert check_cartan_decomp(k, m)

g = k + m

adj = qml.structure_constants(g)

_, k, mtilde, h, new_adj = cartan_subalgebra(g, k, m, adj, start_idx=0)
assert len(h) == 2
assert len(mtilde) == 2
assert len(h) + len(mtilde) == len(m)

new_adj_re = qml.structure_constants(g)

assert np.allclose(new_adj_re, new_adj)
```

Reply:
my new solution uses `einsum` again, but in three separate contractions to avoid huge tensors and for the purpose of doing the moving of the axes automatically

Reply:
So I think this was fine "axes"-wise: In each step, the original `ad` tensor's first axis is contracted with the first axis of `basis_change`, appending the second axis of `basis_change` as the last axis. However, as we do this for all axes of `ad`, these appended axes end up being the _only_ axes of the output, and as they were appended in the right order, all is fine.

The bug was that I used `ad` as input to lines 2 and 3, rather than `new_adj` :facepalm:

Reply:
I like your solution with `einsum` as well and I don't see a reason why that should be slower, so happy to stick to that of course :)